modules/dreamview/frontend/src/renderer/traffic_control.js

```diff
@@ -0,0 +1,270 @@
+            return null;
+        }
+    }
+
+    getStopSignPositionAndHeading(stopSign, coordinates) {
+        const heading = this.getHeadingFromStopLine(stopSign);
+
+        if (!isNaN(heading)) {
+            const stopLinePoint = _.last(stopSign.stopLine[0].segment[0].lineSegment.point);
+            let position = new THREE.Vector3(stopLinePoint.x, stopLinePoint.y, 0);
+            position = coordinates.applyOffset(position);
+
+            return { "pos": position, "heading": heading - Math.PI / 2 };
+        } else {
+            console.error('Error loading stop sign. Unable to determine heading.');
+            return null;
+        }
+    }
+
+    addTrafficControl(items, coordinates, scene) {
+        if (!items || items.length === 0) {
+            return;
+        }
+
+        items.forEach((item) => {
+            let material = null;
+            let object = null;
+            let scales = null;
+            let heading = null;
+            let position = null;
+            let posAndHeading = null;
+            switch (item.type) {
+                case TRAFFIC_LIGHT:
+                    material = trafficLightMaterial;
+                    object = trafficLightObject;
+                    scales = trafficLightScales;
+                    posAndHeading = !OFFLINE_PLAYBACK
+                        && this.getSignalPositionAndHeading(item, coordinates);
+                    break;
+                case STOP_SIGN:
+                    material = stopSignMaterial;
+                    object = stopSignObject;
+                    scales = stopSignScales;
+                    posAndHeading = !OFFLINE_PLAYBACK
+                        && this.getStopSignPositionAndHeading(item, coordinates);
+                    break;
+                default:
+                    break;
+            }
+            if (OFFLINE_PLAYBACK) {
+                heading = item.heading;
+                position = coordinates.applyOffset({
+                    x: item.x,
+                    y: item.y,
+                    z: 0,
+                });
+            } else {
+                heading = posAndHeading.heading;
+                position = posAndHeading.pos;
+            }
+
+            if (heading && position) {
+                loadObject(material, object, scales, mesh => {
+                    mesh.rotation.x = Math.PI / 2;
+                    mesh.rotation.y = heading;
+                    mesh.position.set(position.x, position.y, 0);
+                    mesh.matrixAutoUpdate = false;
+                    mesh.updateMatrix();
+                    scene.add(mesh);
+
+                    const id = OFFLINE_PLAYBACK ? item.id : item.id.id;
+                    this.trafficControl[item.type][id] = mesh;
+                });
+            }
+        });
+    }
+
+    removeTrafficControl(currentItemIds, scene) {
+        Object.keys(currentItemIds)
+            .filter(type => [TRAFFIC_LIGHT, STOP_SIGN].includes(type))
```

Comment:
This is pretty confusing to iterate through the currentItemIds which looks like an array but actually a map keyed by type to extract out ids with specific types. Seems better to take the needed ids out prior to calling the function.